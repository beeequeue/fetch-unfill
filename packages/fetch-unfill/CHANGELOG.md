# fetch-unfill

## 0.2.0

### Minor Changes

- [`c7b9331`](https://github.com/beeequeue/fetch-unfill/commit/c7b9331b91e3174e89c85f85297bcc41ac34e1ce) Thanks [@beeequeue](https://github.com/beeequeue)! - Added more export conditions, e.g. `bun`, `deno`

- [`76f4acc`](https://github.com/beeequeue/fetch-unfill/commit/76f4acc7961482df342050b82a81a02cab1324a3) Thanks [@beeequeue](https://github.com/beeequeue)! - Added `fetch-unfill/aliases` entrypoint that exports a working alias config

## 0.1.0

### Minor Changes

- [`96f0f12`](https://github.com/beeequeue/fetch-unfill/commit/96f0f12c90f61dded0c010d4f64657b363fbbd3c) Thanks [@beeequeue](https://github.com/beeequeue)! - Initial release.
