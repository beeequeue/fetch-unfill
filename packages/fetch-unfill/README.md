See https://github.com/beeequeue/fetch-unfill#readme

---

Install `fetch-unfill` with your package manager, then add it to your bundler's alias config:

```ts
import fetchUnfillAliases from "fetch-unfill/aliases"

export const config = {
  resolve: {
    alias: {
      ...fetchUnfillAliases,
    },
  },
}
```
