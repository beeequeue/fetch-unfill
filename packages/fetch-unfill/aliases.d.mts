type DefaultAliases = {
  "node-fetch": "fetch-unfill"
  "node-fetch-native": "fetch-unfill"
  "cross-fetch": "fetch-unfill"
}

export const rollupAliases: Array<{
  find: keyof DefaultAliases
  replacement: "fetch-unfill"
}>

const fetchUnfillAliases: DefaultAliases
export default fetchUnfillAliases
