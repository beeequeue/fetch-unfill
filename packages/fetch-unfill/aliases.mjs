const aliases = {
  "node-fetch": "fetch-unfill",
  "node-fetch-native": "fetch-unfill",
  "cross-fetch": "fetch-unfill",
}

// prettier-ignore
export const rollupAliases = Object.entries(aliases).map(([find, replacement]) => ({ find, replacement }))

export default aliases
