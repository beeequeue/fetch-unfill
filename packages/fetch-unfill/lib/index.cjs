const fetchFn = window.fetch

window.fetch.polyfill = false
window.polyfill = false

module.exports = fetchFn
Object.assign(module.exports, {
  fetch: fetchFn,
  Request: window.Request,
  Response: window.Response,
  Headers: window.Headers,
  AbortController: window.AbortController,
})
