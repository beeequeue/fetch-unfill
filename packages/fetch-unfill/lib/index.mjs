const fetchFn = window.fetch

window.fetch.polyfill = false
fetchFn.polyfill = false

export default fetchFn

const { Request, Response, Headers, AbortController } = window

export { AbortController, fetchFn as fetch, Headers, Request, Response }
