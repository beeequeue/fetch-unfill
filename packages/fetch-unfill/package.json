{
  "name": "fetch-unfill",
  "type": "module",
  "description": "",
  "version": "0.0.0",
  "license": "MIT",
  "author": "beequeue <adam@haglund.dev>",
  "repository": {
    "type": "github",
    "url": "https://github.com/beeequeue/tinywhich",
    "directory": "./packages/fetch-unfill"
  },
  "engines": {
    "node": ">=18"
  },
  "files": [
    "lib",
    "CHANGELOG.md"
  ],
  "main": "./lib/node.cjs",
  "module": "./lib/node.mjs",
  "browser": "./lib/index.mjs",
  "types": "./lib/index.d.mts",
  "exports": {
    "require": {
      "types": "./lib/index.d.mts",
      "browser": "./lib/index.cjs",
      "default": "./lib/node.cjs"
    },
    "default": {
      "types": "./lib/index.d.mts",
      "browser": "./lib/index.mjs",
      "default": "./lib/node.mjs"
    }
  },
  "scripts": {
    "lint": "eslint"
  }
}
