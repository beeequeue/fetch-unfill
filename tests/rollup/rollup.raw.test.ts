import path from "node:path"

import Alias from "@rollup/plugin-alias"
import CommonJs from "@rollup/plugin-commonjs"
import Json from "@rollup/plugin-json"
import { nodeResolve } from "@rollup/plugin-node-resolve"
import { rollupAliases } from "fetch-unfill/aliases"
import { rollup, type RollupOptions } from "rollup"
import { minify } from "rollup-plugin-esbuild"
import { describe, expect, it } from "vitest"

import { createTester } from "../utils.js"

const test = createTester("rollup", async (name: string, useAlias: boolean = false) => {
  const options = {
    logLevel: "silent",
    plugins: [
      // @ts-expect-error: Incorrect default export
      Json({ compact: true, preferConst: true }),
      // @ts-expect-error: Incorrect default export
      CommonJs(),
      nodeResolve({
        modulePaths: [path.resolve(__dirname, "..", "node_modules")],
        preferBuiltins: true,
      }),
      useAlias
        ? // @ts-expect-error: Incorrect default export
          Alias({
            entries: rollupAliases,
          })
        : null,
      minify(),
    ],
  } satisfies RollupOptions

  const cjs = await rollup({
    ...options,
    input: path.resolve(__dirname, "..", "fixtures", `${name}.cjs`),
  }).then(async (c) =>
    c.generate({
      file: "cjs.js",
      format: "cjs",
      compact: true,
      inlineDynamicImports: true,
    }),
  )

  const mjs = await rollup({
    ...options,
    input: path.resolve(__dirname, "..", "fixtures", `${name}.mjs`),
  }).then(async (c) =>
    c.generate({
      file: "mjs.js",
      format: "esm",
      compact: true,
      inlineDynamicImports: true,
    }),
  )

  expect(cjs.output).toHaveLength(1)
  expect(mjs.output).toHaveLength(1)

  return {
    cjs: cjs.output[0].code,
    mjs: mjs.output[0].code,
  }
})

describe("node-fetch", () => {
  it("bundles the original package", async () => {
    await test("node-fetch", false)
  })

  it("unfills it", async () => {
    await test("node-fetch", true, true)
  })
})

describe("cross-fetch", () => {
  it("bundles the original package", async () => {
    await test("cross-fetch", false)
  })

  it("unfills it", async () => {
    await test("cross-fetch", true, true)
  })
})
