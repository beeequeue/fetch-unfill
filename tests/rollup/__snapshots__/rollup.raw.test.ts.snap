// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cross-fetch > bundles the original package > cjs 1`] = `"302.85KB"`;

exports[`cross-fetch > bundles the original package > mjs 1`] = `"302.75KB"`;

exports[`cross-fetch > unfills it > cjs 1`] = `"640B"`;

exports[`cross-fetch > unfills it > mjs 1`] = `"175B"`;

exports[`cross-fetch > unfills it > unfilled cjs code 1`] = `
""use strict";function a(e){return e&&e.__esModule&&Object.prototype.hasOwnProperty.call(e,"default")?e.default:e}var t={exports:{}},s={exports:{}},o;function c(){return o||(o=1,function(e){const r=globalThis.fetch;globalThis.fetch.polyfill=!1,r.polyfill=!1,e.exports=r,Object.assign(e.exports,{fetch:r,Request:globalThis.Request,Response:globalThis.Response,Headers:globalThis.Headers,AbortController:globalThis.AbortController})}(s)),s.exports}var l;function p(){if(l)return t.exports;l=1;const e=c();return t.exports=()=>e("https://example.com"),t.exports.example=()=>e("https://example.com"),t.exports}var n=p(),h=a(n);module.exports=h;
"
`;

exports[`cross-fetch > unfills it > unfilled mjs code 1`] = `
"const l=globalThis.fetch;globalThis.fetch.polyfill=!1,l.polyfill=!1;var e=()=>l("https://example.com");const t=()=>l("https://example.com");export{e as default,t as example};
"
`;

exports[`node-fetch > bundles the original package > cjs 1`] = `"91.89KB"`;

exports[`node-fetch > bundles the original package > mjs 1`] = `"89.91KB"`;

exports[`node-fetch > unfills it > cjs 1`] = `"640B"`;

exports[`node-fetch > unfills it > mjs 1`] = `"175B"`;

exports[`node-fetch > unfills it > unfilled cjs code 1`] = `
""use strict";function a(e){return e&&e.__esModule&&Object.prototype.hasOwnProperty.call(e,"default")?e.default:e}var t={exports:{}},r={exports:{}},s;function n(){return s||(s=1,function(e){const o=globalThis.fetch;globalThis.fetch.polyfill=!1,o.polyfill=!1,e.exports=o,Object.assign(e.exports,{fetch:o,Request:globalThis.Request,Response:globalThis.Response,Headers:globalThis.Headers,AbortController:globalThis.AbortController})}(r)),r.exports}var l;function p(){if(l)return t.exports;l=1;const e=n();return t.exports=()=>e("https://example.com"),t.exports.example=()=>e("https://example.com"),t.exports}var c=p(),h=a(c);module.exports=h;
"
`;

exports[`node-fetch > unfills it > unfilled mjs code 1`] = `
"const l=globalThis.fetch;globalThis.fetch.polyfill=!1,l.polyfill=!1;var e=()=>l("https://example.com");const t=()=>l("https://example.com");export{e as default,t as example};
"
`;
